### PR TITLE
KV Store: Implement settings manager

### LINKED ISSUE #3785: KV Settings Manager
SettingsManager relies on BranchLocker to keep the object store settings file consistent. 
Since we no longer have a `BranchLocker` in the kv world, we need to replace the settings manager.

See previous related discussions: #2491 #2498

### PR DESCRIPTION
**Pull request by N-o-Z**
Implemented settings manager in DB only
Added blockstore access in migration flow and verified it works
Closes #3785 #3830

### CODE REVIEW

File: pkg/graveler/settings/manager.go [hunk 1]
```diff
@@ -110,54 +125,51 @@ func (m *Manager) Get(ctx context.Context, repositoryID graveler.RepositoryID, k
...
+		err = m.store.SetMsgIf(ctx, graveler.RepoPartition(&graveler.RepositoryRecord{RepositoryID: repositoryID, Repository: repo}), []byte(graveler.SettingsPath(key)), data, pred)
+		if errors.Is(err, kv.ErrPredicateFailed) {
+			err = graveler.ErrPreconditionFailed
+		} else if err != nil {
+			return backoff.Permanent(err)
+		}
+		return err
+	}, bo)
+	if bo.NextBackOff() == bo.Stop {
+		return fmt.Errorf("update settings: %w", graveler.ErrTooManyTries)
```
Comment (itaiad200) [inline pkg/graveler/settings/manager.go:170]: Do we log the errors that occur? If we maxed our retries with outputing any indication for the reason then it's hard for the user/developer to understand what went wrong..

File: pkg/graveler/settings/manager_test.go [hunk 2]
```diff
@@ -95,7 +193,7 @@ func TestUpdateWithLock(t *testing.T) {
 	wg.Add(IncrementCount)
 	for i := 0; i < IncrementCount; i++ {
 		go func() {
-			testutil.Must(t, m.UpdateWithLock(ctx, "example-repo", "settingKey", emptySettings, update))
+			testutil.Must(t, m.Update(ctx, "example-repo", "settingKey", emptySettings, update))
 			wg.Done()
 		}()
```
Comment (itaiad200) [inline pkg/graveler/settings/manager_test.go:198]: This is flaky by nature in the KV implementation. There's a chance that a certain goroutine will fail to update the settings within the number of retries allocated to it.

File: pkg/graveler/settings/manager_test.go [hunk 3]
```diff
@@ -36,45 +40,139 @@ func (m *mockCache) GetOrSet(k interface{}, setFn cache.SetFn) (v interface{}, e
...
+		return testErr
+	}
+	require.ErrorIs(t, m.Update(ctx, "example-repo", "settingKey", emptySettings, update), testErr)
+	gotSettings, err = m.GetLatest(ctx, "example-repo", "settingKey", emptySettings)
+	if diff := deep.Equal(validationData, gotSettings); diff != nil {
 		t.Fatal("got unexpected settings:", diff)
 	}
 }
 
+// TODO: locking is irrelevant for KV, create a new test for it
```
Comment (itaiad200) [inline pkg/graveler/settings/manager_test.go:166]: Let's add this now instead of this flaky test

**Review by itaiad200 — APPROVED** [review]
Looks great! Approving but consider the following:
1. Introduced a flaky test, please use a dedicated one for the kv impl.
2. Add settings to the migration test in esti

**Review by itaiad200 — APPROVED** [review]

**Comment by N-o-Z** [comment]
@itaiad200 Please review again with changes to Mem Adapter and MigrateTest

File: pkg/block/mem/adapter.go [hunk 4]
```diff
@@ -63,11 +64,11 @@ type Adapter struct {
 }
 
 // FetchStore Returns mem store if already initialized, otherwise creates one
-func FetchStore(opts ...func(a *Adapter)) *Adapter {
-	if memStore != nil {
+func FetchStore(params params.Mem) *Adapter {
```
Comment (itaiad200) [inline pkg/block/mem/adapter.go:67]: Why would I call `FetchStore` with `params.ReuseStore=false`? I think it makes this parameter or this function redundant.

File: pkg/config/template.go [hunk 5]
```diff
@@ -138,6 +138,9 @@ type configuration struct {
 		Local                  *struct {
 			Path string
 		}
+		Mem *struct {
+			ReuseStore bool `mapstructure:"reuse_store"`
```
Comment (itaiad200) [inline pkg/config/template.go:142]: It's an in-memory test only implementation, I don't understand why we would like to add this flag to the configuration of lakeFS

File: pkg/graveler/settings/manager.go [hunk 6]
```diff
@@ -160,6 +160,7 @@ func (m *KVManager) Update(ctx context.Context, repositoryID graveler.Repository
 		}
 		err = m.store.SetMsgIf(ctx, graveler.RepoPartition(&graveler.RepositoryRecord{RepositoryID: repositoryID, Repository: repo}), []byte(graveler.SettingsPath(key)), data, pred)
 		if errors.Is(err, kv.ErrPredicateFailed) {
+			logging.Default().WithError(err).Warn("Predicate failed on settings update. Retrying")
```
Comment (itaiad200) [inline pkg/graveler/settings/manager.go:163]: Warning is preferred and not an error, since this is expected for some usages of the settings manager.

**Review by itaiad200 — CHANGES_REQUESTED** [review]

**Review by itaiad200 — APPROVED** [review]
LGTM 💪 🍄 🎹 🎵

**Pull request merged**